tools/oprt2/pkg/packagemanager/apt/apt.go

```diff
@@ -0,0 +1,243 @@
+/*
+ *  Copyright 2025 Gravitational, Inc
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package apt
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"maps"
+	"regexp"
+	"slices"
+
+	"github.com/gravitational/shared-workflows/tools/oprt2/pkg/commandrunner"
+	"github.com/gravitational/shared-workflows/tools/oprt2/pkg/filemanager"
+	"github.com/gravitational/shared-workflows/tools/oprt2/pkg/logging"
+	"github.com/gravitational/shared-workflows/tools/oprt2/pkg/packagemanager"
+)
+
+// APT is a [packagemanager.Manager] that manages Debian files to APT repos.
+type APT struct {
+	components   map[string][]*regexp.Regexp
+	distros      map[string][]string
+	fm           filemanager.FileManager
+	hooks        []commandrunner.Hook
+	attuneRunner *commandrunner.Runner
+	logger       *slog.Logger
+}
+
+var _ packagemanager.Manager = (*APT)(nil)
+
+// APTOption provides optional configuration to the APT package manager.
+type APTOption func(apt *APT)
+
+// WithAttuneHooks adds hooks to the Attune command runner.
+func WithAttuneHooks(hooks ...commandrunner.Hook) APTOption {
+	return func(apt *APT) {
+		apt.hooks = append(apt.hooks, hooks...)
+	}
+}
+
+// WithComponents configures the package manager to map files to APT components. Key is
+// the component name, value is a list of regular expressions used to match files to the
+// component. Each expression will be matched against every file in the storage backend,
+// including the file's path. For example, `some/path/teleport.deb`.
+func WithComponents(components map[string][]*regexp.Regexp) APTOption {
+	return func(apt *APT) {
+		if len(apt.components) == 0 {
+			apt.components = make(map[string][]*regexp.Regexp, len(components))
+		}
+
+		// Deep merge the maps
+		for componentName, fileNameMatchers := range components {
+			if _, ok := apt.components[componentName]; !ok {
+				components[componentName] = fileNameMatchers
+				continue
+			}
+			components[componentName] = append(apt.components[componentName], fileNameMatchers...)
+		}
```

Comment:
I get that it tries to merge `APT`'s internal `components` field with components from the functional argument but my question is why would this happen? The "New" method below doesn't set `.components` field on the object so they can only be set via this functional argument.

The only way this logic would be useful is if you provide `WithComponents` option multiple times but that seems unnecessary complication. Can this be simplified to just take components provided without trying to merge?

Reply:
Sorry I think I conflated this with one of the other comments. Yea instead of merging the maps here and below, I can just have it be overwritten, and document this behavior with a godoc. That said the uniqueness collection stuff needs to stay.